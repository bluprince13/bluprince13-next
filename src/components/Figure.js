const Reference = ({ reference, referenceHref }) => (
    <span>
        {' '}
        {referenceHref ? (
            <a href={referenceHref}>{`[${reference}]`}</a>
        ) : (
            [reference]
        )}
    </span>
)

const Caption = ({ caption, href, reference, referenceHref }) => (
    <figcaption>
        {href ? <a href={href}>{caption}</a> : caption}
        <Reference reference={reference} referenceHref={referenceHref} />
    </figcaption>
)

const getMaxWidth = (size) => {
    switch (size) {
        case 's':
            return '200px'
        case 'l':
            return ''
        case 'm':
        default:
            return '350px'
    }
}

// TODO: Use https://nextjs.org/docs/api-reference/next/image when it's able to
// infer width and height automatically.
// See https://github.com/vercel/next.js/issues/18497
const Figure = ({
    src,
    alt,
    caption,
    href,
    reference,
    referenceHref,
    maxWidth,
    size = 'm'
}) => {
    const maxWidthToUse = maxWidth || getMaxWidth(size)

    return (
        <figure>
            <img
                src={src}
                alt={alt ?? caption}
                width="100%"
                style={{ maxWidth: maxWidthToUse }}
            />
            {caption ? (
                <Caption
                    caption={caption}
                    href={href}
                    reference={reference}
                    referenceHref={referenceHref}
                />
            ) : (
                ''
            )}
        </figure>
    )
}

export default Figure
