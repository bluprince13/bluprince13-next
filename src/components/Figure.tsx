/* eslint-disable react/jsx-filename-extension */
/* eslint-disable @next/next/no-img-element */
const Reference = ({
    reference,
    referenceHref
}: {
    reference: string
    referenceHref: string
}) => (
    <span>
        {' '}
        {referenceHref ? (
            <a href={referenceHref}>{`[${reference}]`}</a>
        ) : (
            [reference]
        )}
    </span>
)

const Caption = ({ caption, href, reference, referenceHref }) => (
    <figcaption>
        {href ? <a href={href}>{caption}</a> : caption}
        <Reference reference={reference} referenceHref={referenceHref} />
    </figcaption>
)

const getMaxWidth = (size) => {
    switch (size) {
        case 'xs':
            return '75px'
        case 's':
            return '200px'
        case 'ml':
            return '500px'
        case 'l':
            return ''
        case 'm':
        default:
            return '350px'
    }
}

// TODO: Use https://nextjs.org/docs/api-reference/next/image when it's able to
// infer width and height automatically.
// See https://github.com/vercel/next.js/issues/18497
type FigureProps = {
    src: string
    alt?: string
    caption?: string
    href?: string
    reference?: string
    referenceHref?: string
    maxWidth?: string
    size?: 'xs' | 's' | 'm' | 'ml' | 'l'
    align?: 'center'
}

const Figure = ({
    src,
    alt,
    caption,
    href,
    reference,
    referenceHref,
    maxWidth,
    size = 'm',
    align
}: FigureProps) => {
    const maxWidthToUse = maxWidth || getMaxWidth(size)
    const style: { margin?: string; display?: string; maxWidth?: string } = {
        maxWidth: maxWidthToUse
    }
    if (align == 'center') {
        style.margin = 'auto'
        style.display = 'block'
    }
    return (
        <figure style={{ marginLeft: 0, marginRight: 0 }}>
            <img src={src} alt={alt ?? caption} width="100%" style={style} />
            {caption ? (
                <Caption
                    caption={caption}
                    href={href}
                    reference={reference}
                    referenceHref={referenceHref}
                />
            ) : (
                ''
            )}
        </figure>
    )
}

export default Figure
