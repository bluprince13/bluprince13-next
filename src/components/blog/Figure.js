export default ({ src, alt, caption }) => (
    <figure>
        <img src={src} alt={alt} width="100%" style={{ maxWidth: '500px' }} />
        <figcaption>{caption}</figcaption>
    </figure>
)
