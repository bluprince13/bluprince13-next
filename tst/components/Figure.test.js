import Figure from '@Components/Figure'
import { cleanup, render, screen } from '@testing-library/react'

describe('Figure', () => {
    const setup = ({ size = 'm', maxWidth } = {}) => {
        const queries = render(
            <Figure
                src="https://src.com"
                alt="alt"
                caption="caption"
                href="https://href.com"
                reference="reference"
                referenceHref="https://reference-href.com"
                maxWidth={maxWidth}
                size={size}
            />
        )
        const image = screen.getByAltText('alt')

        return { image, ...queries }
    }

    it('renders correctly', () => {
        const { image } = setup()

        const caption = screen.getByText('caption')
        const reference = screen.getByText('[reference]')
        expect(caption.href).toBe('https://href.com/')
        expect(reference.href).toBe('https://reference-href.com/')
        expect(image.src).toBe('https://src.com/')
        expect(image).toHaveStyle('max-width: 350px')

        // TODO: Remove snapshot and test image width directly
        expect(image).toMatchSnapshot()
    })

    it('sets correct max-width for given size', () => {
        const { image: small } = setup({ size: 's' })
        expect(small).toHaveStyle('max-width: 200px')
        cleanup()

        const { image: medium } = setup({ size: 'm' })
        expect(medium).toHaveStyle('max-width: 350px')
        cleanup()

        const { image: large } = setup({ size: 'l' })
        expect(large).toHaveStyle('max-width:')
    })

    it('sets correct max-width for specified max-width', () => {
        const { image } = setup({ maxWidth: '13px' })

        expect(image).toHaveStyle('max-width: 13px')
    })
})
