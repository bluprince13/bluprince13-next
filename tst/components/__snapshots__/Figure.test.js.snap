// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Figure renders correctly 1`] = `
<DocumentFragment>
  <figure
    style="margin-left: 0px; margin-right: 0px;"
  >
    <img
      alt="alt"
      src="https://src.com"
      style="max-width: 350px;"
      width="100%"
    />
    <figcaption>
      <a
        href="https://href.com"
      >
        caption
      </a>
      <span>
         
        <a
          href="https://reference-href.com"
        >
          [reference]
        </a>
      </span>
    </figcaption>
  </figure>
</DocumentFragment>
`;
